github/repos.go

```diff
@@ -1719,42 +1729,216 @@ func (s *RepositoriesService) ReplaceAppRestrictions(ctx context.Context, owner,
+	}
+
+	var teams []*Team
+	resp, err := s.client.Do(ctx, req, &teams)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return teams, resp, nil
+}
+
+// ReplaceTeamRestrictions replaces the team that have push access to a given protected branch.
+// This removes all teams that previously had push access and grants push access to the new list of teams.
+// It requires the GitHub teams to have `write` access to the `content` permission.
+//
+// Note: The list of users, apps, and teams in total is limited to 100 items.
+//
+// GitHub API docs: https://docs.github.com/en/rest/branches/branch-protection#set-team-access-restrictions
+func (s *RepositoriesService) ReplaceTeamRestrictions(ctx context.Context, owner, repo, branch string, teams []string) ([]*Team, *Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/branches/%v/protection/restrictions/teams", owner, repo, branch)
+	req, err := s.client.NewRequest("PUT", u, teams)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var newTeams []*Team
+	resp, err := s.client.Do(ctx, req, &newTeams)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return newTeams, resp, nil
+}
+
+// AddTeamRestrictions grants the specified teams push access to a given protected branch.
+// It requires the GitHub teams to have `write` access to the `content` permission.
+//
+// Note: The list of users, apps, and teams in total is limited to 100 items.
+//
+// GitHub API docs: https://docs.github.com/en/rest/branches/branch-protection#add-team-access-restrictions
+func (s *RepositoriesService) AddTeamRestrictions(ctx context.Context, owner, repo, branch string, teams []string) ([]*Team, *Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/branches/%v/protection/restrictions/teams", owner, repo, branch)
+	req, err := s.client.NewRequest("POST", u, teams)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var newTeams []*Team
+	resp, err := s.client.Do(ctx, req, &newTeams)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return newTeams, resp, nil
+}
+
+// RemoveTeamRestrictions removes the restrictions of a team from pushing to this branch.
+// It requires the GitHub teams to have `write` access to the `content` permission.
+//
+// Note: The list of users, apps, and teams in total is limited to 100 items.
+//
+// GitHub API docs: https://docs.github.com/en/rest/branches/branch-protection#remove-team-access-restrictions
+func (s *RepositoriesService) RemoveTeamRestrictions(ctx context.Context, owner, repo, branch string, teams []string) ([]*Team, *Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/branches/%v/protection/restrictions/teams", owner, repo, branch)
+	req, err := s.client.NewRequest("DELETE", u, teams)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var newTeams []*Team
+	resp, err := s.client.Do(ctx, req, &newTeams)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return newTeams, resp, nil
+}
+
+// ListUserRestrictions lists the GitHub users that have push access to a given protected branch.
```

Comment:
Let's please change this to `ListUsersRestrictions` , since we are returning potentially many users and not a single user.

```suggestion
// ListUsersRestrictions lists the GitHub users that have push access to a given protected branch.
```

Reply:
I see.  OK, let's leave this as well, then.